### PR TITLE
Initial setup of Prettier & ESLint

### LINKED ISSUE #884: Setup Linting for the project
We want to unify our style and formatting so we run into fewer merge conflicts. Whatever we use should work with VScode (and other editors) so developers can automatically fix the style in the code as they're working on it. It also needs to run in CI so CI can enforce our code style and format.

We'd also like to do a single-commit change that updates the style in the repo to the new standard. This commit should be added to a `.git-blame-ignore-revs` file so we keep the `blame` command useful.

It looks like Meteor recommends utilizing `eslint` with the `airbnb` style guide. We'll go ahead and add that configuration so that we're writing code in the way Meteor recommends.

### PR DESCRIPTION
**Pull request by ian-noaa**
We're using Prettier to autoformat our code. ESLint will be used to check for code quality. Currently, I've changed the ESLint checks that can't be automatically fixed to warnings.

I've also set up GitHub actions to fail PRs that fail the checks from Prettier and ESLint.

Closes #884

### CODE REVIEW

**Comment by mollybsmith-noaa** [comment]
Can you do this for the METexpress apps as well? It’s good to keep all the apps with the same settings.

**Comment by ian-noaa** [comment]
Yup! I was thinking I'd get this setup on MATS and then roll it out to METexpress and MATScommon once we were satisfied.

File: .prettierrc
Comment (ian-noaa) [inline .prettierrc:2]: I ended up going with a line length of 88 here as that's what's recommended by Python's Black formatter and it seems nice to have a consistent line length between Python and JS. Generally, keeping line lengths between 60-100 chars is recommended for readability. I'm open to changing this within that range if anyone has strong opinions.

https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length

**Comment by ian-noaa** [comment]
Hey @mollybsmith-noaa, @randytpierce, & @gopa-noaa,

I'd appreciate your review of this. I've gone ahead with Prettier for formatting and ESLint for code-quality checks - using Airbnb's checks as [Meteor recommended](https://guide.meteor.com/code-style.html#eslint). The big changes are 9b8f07b (reformat with Prettier) and 78b12e0 (run ESLint's automated fixer on the codebase).

Both tools have assured me that there shouldn't be bugs. I'm slightly more skeptical of ESLint's claims so if anyone wants to do a quick smoke test, I'd appreciate it. 

My biggest quibble with how this is configured is the explosion of `.prettier` and `.eslintrc.json` files in the code base. This is a limitation of ESLint's config - they recently released a [new config system](https://eslint.org/blog/2022/08/new-config-system-part-1/) that we might be able to use to consolidate our config into one place. However, it's still pretty young so for now the old config system seemed like the least-bad route.

**Comment by ian-noaa** [comment]
Mmmm... I forgot to mention - the easiest way to see the changes might be to checkout the `setup-eslint` branch locally.

File: .prettierrc
Comment (mollybsmith-noaa) [reply .prettierrc:2]: Sigh, the 100 was already kind of narrow. Well, I guess we can do it if it's what the rest of the world wants.

**Comment by mollybsmith-noaa** [comment]
FYI, there are going to be things that break if prettier changes all strings to have single quotes. In order to be consistent with JSON, which is used quite a bit in MATS, we've been having double-quoted strings _which contain single quotes as part of the string_. This occurs all over the place, especially in the app data routines, and I'm worried that if we change it, we'll mess up our JSON parsing.

**Comment by mollybsmith-noaa** [comment]
> FYI, there are going to be things that break if prettier changes all strings to have single quotes. In order to be consistent with JSON, which is used quite a bit in MATS, we've been having double-quoted strings _which contain single quotes as part of the string_. This occurs all over the place, especially in the app data routines, and I'm worried that if we change it, we'll mess up our JSON parsing.

For example, `statisticClause = "sum(m0.yy) as hit, sum(m0.yn) as fa, sum(m0.ny) as miss, sum(m0.nn) as cn, group_concat(m0.time, ';', m0.yy, ';', m0.yn, ';', m0.ny, ';', m0.nn order by m0.time) as sub_data, count(m0.yy) as N0";`
If this string is single-quoted with double quotes as part of the string, then the JSON parser sticks lots of escapes into the string, which other parts of the application don't like.

**Comment by mollybsmith-noaa** [comment]
> > FYI, there are going to be things that break if prettier changes all strings to have single quotes. In order to be consistent with JSON, which is used quite a bit in MATS, we've been having double-quoted strings _which contain single quotes as part of the string_. This occurs all over the place, especially in the app data routines, and I'm worried that if we change it, we'll mess up our JSON parsing.
> 
> For example, `statisticClause = "sum(m0.yy) as hit, sum(m0.yn) as fa, sum(m0.ny) as miss, sum(m0.nn) as cn, group_concat(m0.time, ';', m0.yy, ';', m0.yn, ';', m0.ny, ';', m0.nn order by m0.time) as sub_data, count(m0.yy) as N0";` If this string is single-quoted with double quotes as part of the string, then the JSON parser sticks lots of escapes into the string, which other parts of the application don't like.

Although it looks like that particular line was left alone, so maybe it is being smart about things.

**Comment by ian-noaa** [comment]
Oh, hah. That one's on me. I thought it was the other way around and we wanted single quotes instead of double quotes. Prettier by default uses double quotes everywhere so I switched it in the config. That'll be easy to change back. 🙂 

https://github.com/NOAA-GSL/MATS/blob/8a43d7af789f8e01c6cc1e65f78e5e4c4f4b8764/.prettierrc#L3

**Comment by mollybsmith-noaa** [comment]
Honestly it wouldn't be a problem if we didn't use so much JSON everywhere! :) I think the airbnb style guide does want single quotes, but the two are at odds.

**Review by mollybsmith-noaa — COMMENTED** [review]
Ian, these changes tentatively look good, but they’re _really_ extensive. The linter has substantially changed a lot of code. We need to run the automated tests on every app before merging this. I can do this, or you can?

**Comment by ian-noaa** [comment]
I switched Prettier back to double quotes. LMK what you think.

And agreed. That's always the scary thing about turning linters on - they touch everything. The linters both promised the automated fixes were safe. I feel like I can trust Prettier on that - it reads the code into an AST (Abstract Syntax Tree) and spits it back out so that the AST is the same. However, I'm a bit more skeptical of ESLint's claims since it can do things that change the AST. 

I've been trying to get my machine set up to do some manual testing here. However, given the extent of the changes here, I'd really appreciate it if you could run the automated tests over the apps.

**Comment by mollybsmith-noaa** [comment]
Will do. One thing of low importance, which may be too late anyway--can we configure the linter not to do this: `const { diffFrom } = curve;`, instead of `const diffFrom = curve.diffFrom;`? The first way has reached perl levels of unreadability, although maybe I'm just yelling at kids to get off my lawn.

**Comment by mollybsmith-noaa** [comment]
We should revert the cb-ceiling app until Gopa is finished with the new data routines, or that's going to be one hell of a merge conflict. All of the automated tests are passing so far, though!

**Comment by gopa-noaa** [comment]
No luck so far on braces, I guess.  No a deal breaker, I will re-train my eyes.

**Review by mollybsmith-noaa — APPROVED** [review]
All the tests pass, so I approve this merge, but as stated before, we should probably hold off formatting cb-ceiling files until Gopa is done. Nice work, though!

**Review by gopa-noaa — APPROVED** [review]

**Comment by ian-noaa** [comment]
Thanks all, I'll look into ways to avoid a merge conflict with Gopa's changes. Gopa - my apologies on the braces. 😞 

Molly - I agree that the following looks weird to me as well. (This example from `apps/anomalycor/server/dataFunctions/data_dieoff.jsL45-46`)

```diff
...
    const curve = curves[curveIndex];
-    const diffFrom = curve.diffFrom;
+    const { diffFrom } = curve;
-    const label = curve.label;
+    const { label } = curve;
...
```

It looks like the change is from ESLint's [`prefer-destructuring`](https://eslint.org/docs/latest/rules/prefer-destructuring) rule. [MDN's docs on destructuring assignment]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) are helpful for seeing how the JS language folk are imagining the feature's usage.

Given the documentation there, I suspect this is a limit of ESLint's automated "fixing" and it'd be better rewritten as something like:

```diff
...
    const curve = curves[curveIndex];
-    const diffFrom = curve.diffFrom;
-    const label = curve.label;
+    const { label, diffFrom } = curve;
...
```

Or, looking at the rest of the `dataDieOff` function:

```diff
...
    const curve = curves[curveIndex];
+    const { label, diffFrom, variable, dieoff-type: forecastLengthStr  } = curve;
-    const diffFrom = curve.diffFrom;
-    const label = curve.label;
...
-    const { variable } = curve;
...
-    const forecastLengthStr = curve["dieoff-type"];
```

Assuming the `curves` object had keys like so: (modified from the MDN docs)

```js
const curve = { label: ..., diffFrom: ... };
const { label, diffFrom } = curve;
// is equivalent to:
// const label = curve.label;
// const diffFrom = curve.diffFrom;
```

It looks like the case where you're pulling a single key out of an object to a differently named variable is allowed by ESLint. (Though ESLint has an option to disable that as well)

```js
const curve = { label: ..., diffFrom: ... };
const label = curve.label;  // ESLint error
const mylabel = curve.label;  // ESLint is okay with this, but there are options to error on this if desired
const { label: mylabel } = curve;  // ESLint is okay with this 
```

I could disable the ESLint rule. However, given that it comes from the [Airbnb guidelines](https://github.com/airbnb/javascript#destructuring) (at least for accessing multiple values) my inclination would be to leave it and see what we think over the longer run/with additional refactoring. If we run into cases where we truly need a single value from an object in a function, I could see decorating the call to disable eslint for that line like so since I agree that the original method is more readable when you're going after a single value: 

```js
// eslint-disable-next-line prefer-destructuring
const diffFrom = curve.diffFrom;
```

**Comment by gopa-noaa** [comment]
Thanks Ian.
Now that I have both formatters in my VSCode, I have set the default to ESLint, but if/when my eyes need a good-old formatted  look ( makes me understand code structure a bit faster) , I can temporarily switch to the other format.  ... so all good from my side.

**Comment by ian-noaa** [comment]
This is going to be hard to merge as-is with other open PRs. It seems better to split this PR into 2-3 parts. 

1. Just the config changes to setup the linters (#889)
2. The application of the linters on the codebase and setting up CI (TBD)
3. Adding the `.git-blame-ignore-revs` file to keep `git blame` useful

This way, open PRs can be rebased on the result of #889, and then apply the linters themselves to reduce merge conflicts.

**Pull request closed without merging**
